.semaphore/semaphore.yml

```diff
@@ -0,0 +1,176 @@
+        commands:
+          - NPROC=$(nproc)
+          - GO_VERSION=1.19.4
+          - PROTOC_VERSION=21.12
+          - PROTOC_GEN_VERSION=v1.28.1
+          - PROTOC_GRPC_VERSION=v1.2.0
+      env_vars: []
+    dependencies:
+      - Setup
+  - name: Lint
+    dependencies:
+      - Setup
+    task:
+      jobs:
+        - name: Linting
+          commands:
+            - |
+              function version_gt() {
+                test "$(printf ''%s\n'' "$@" | sort -V | head -n 1)" != "$1"
+              }
+            - |
+              CURRENT_GO_VERSION=$(go version | sed 's/[^0-9.]*\([0-9.]*\).*/\1/')
+            - |
+              if version_gt $GO_VERSION $CURRENT_GO_VERSION; then
+                echo "Updating go version from $CURRENT_GO_VERSION to $GO_VERSION ..."
+                sudo rm -rf /usr/bin/go
+                sudo rm -rf /usr/local/go
+                sudo rm -rf /usr/local/bin/go
+                sudo rm -rf /usr/local/golang
+                sudo rm -rf $GOROOT $GOPATH $GOBIN
+                wget https://go.dev/dl/go$GO_VERSION.linux-amd64.tar.gz
+                sudo tar -C /usr/local -xzf go$GO_VERSION.linux-amd64.tar.gz
+                export GOROOT=/usr/local/go
+                export PATH=$PATH:/usr/local/go/bin
+              fi
+            - export GOBIN=$HOME/go/bin
+            - 'export PATH=$PATH:$GOBIN'
+            - checkout
+            - 'curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOBIN) $GOLANGCI_LINT_VERSION'
+            - 'golangci-lint run --no-config --disable-all --enable gosimple,errcheck,govet,unused,goconst,godot --timeout 3m'
+      prologue:
+        commands:
+          - '# This script contains commands to be executed by the CI tool.'
+          - NPROC=$(nproc)
+          - GO_VERSION=1.19.4
+          - PROTOC_VERSION=21.12
+          - PROTOC_GEN_VERSION=v1.28.1
+          - PROTOC_GRPC_VERSION=v1.2.0
+          - GOLANGCI_LINT_VERSION=v1.52.1
+  - name: Modified Tests
+    dependencies:
+      - Setup
+    task:
+      jobs:
+        - name: Tests on Modified modules
+          commands:
+            - checkout
+            - echo "Running modified tests...."
+            - echo "" > coverage.txt
+            - |
+              # Get the list of modified Go files
+              MODIFIED_FILES=$(git diff --name-only --diff-filter=AM | grep '\.go$')
+
+              # Extract the unique Go module names from modified files
+              MODIFIED_MODULES=$(go list -e -f '{{.Module.Path}}' $(dirname $MODIFIED_FILES) | sort -u)
+
+              # Run tests for modified modules
+              for module in $MODIFIED_MODULES; do
+                go test -mod=vendor -v -race -tags test -coverprofile=profile.out -covermode=atomic "$module/..."
+                if [ -f profile.out ]; then
+                  cat profile.out >> coverage.txt
+                  rm profile.out
+                fi
+              done     

```

Comment:
We have only one module. Also, detecting changes for testing is more complex than checking changed files, since it depends on **which files are changed**: if we change access control, for example, we need to test messaging, as well.